src/mbgl/gl/context.cpp

```diff
@@ -327,11 +262,189 @@ void Context::setDirtyState() {
+        clearColor = *color;
+        colorMask = { true, true, true, true };
+    }
+
+    if (depth) {
+        mask |= GL_DEPTH_BUFFER_BIT;
+        clearDepth = *depth;
+        depthMask = true;
+    }
+
+    if (stencil) {
+        mask |= GL_STENCIL_BUFFER_BIT;
+        clearStencil = *stencil;
+        stencilMask = 0xFF;
+    }
+
+    MBGL_CHECK_ERROR(glClear(mask));
+}
+
+#if not MBGL_USE_GLES2
+DrawMode Context::operator()(const Points& points) {
+    pointSize = points.pointSize;
+    return DrawMode::Points;
+}
+#else
+DrawMode Context::operator()(const Points&) {
+    return DrawMode::Points;
+}
+#endif // MBGL_USE_GLES2
+
+DrawMode Context::operator()(const Lines& lines) {
+    lineWidth = lines.lineWidth;
+    return DrawMode::Lines;
+}
+
+DrawMode Context::operator()(const LineStrip& lineStrip) {
+    lineWidth = lineStrip.lineWidth;
+    return DrawMode::LineStrip;
+}
+
+DrawMode Context::operator()(const Triangles&) {
+    return DrawMode::Triangles;
+}
+
+DrawMode Context::operator()(const TriangleStrip&) {
+    return DrawMode::TriangleStrip;
+}
+
+std::size_t Context::VertexArrayObjectHash::operator()(const VertexArrayObjectKey& key) const {
+    std::size_t seed = 0;
+    boost::hash_combine(seed, std::get<0>(key));
+    boost::hash_combine(seed, std::get<1>(key));
+    boost::hash_combine(seed, std::get<2>(key));
+    boost::hash_combine(seed, std::get<3>(key));
+    return seed;
+}
+
+void Context::setDepth(const Depth& depth) {
+    if (depth.func == Depth::Always && !depth.mask) {
+        depthTest = false;
+    } else {
+        depthTest = true;
+        depthFunc = depth.func;
+        depthMask = depth.mask;
+        depthRange = depth.range;
+    }
+}
+
+void Context::setStencil(const Stencil& stencil) {
+    if (stencil.test.is<Stencil::Always>() && !stencil.mask) {
+        stencilTest = false;
+    } else {
+        stencilTest = true;
+        stencilMask = stencil.mask;
+        stencilOp = { stencil.fail, stencil.depthFail, stencil.pass };
+        apply_visitor([&] (const auto& test) {
+            stencilFunc = { test.func, stencil.ref, test.mask };
+        }, stencil.test);
+    }
+}
+
+void Context::setColor(const Color& color) {
+    if (color.blendFunction.is<Color::Replace>()) {
+        blend = false;
+    } else {
+        blend = true;
+        blendColor = color.blendColor;
+        apply_visitor([&] (const auto& blendFunction) {
+            // TODO: blendEquation = blendFunction.equation;
+            blendFunc = { blendFunction.srcFactor, blendFunction.dstFactor };
+        }, color.blendFunction);
+    }
+
+    colorMask = color.mask;
+}
+
+void Context::draw(const Drawable& drawable) {
+    DrawMode mode = apply_visitor([&] (auto m) { return (*this)(m); }, drawable.mode);
+
+    setDepth(drawable.depth);
+    setStencil(drawable.stencil);
+    setColor(drawable.color);
+
+    program = drawable.program;
+
+    drawable.bindUniforms();
+
+    for (const auto& segment : drawable.segments) {
```

Comment:
I think it'll be short circuited prior to getting here, but I'll add an internal guard as well.